### PR TITLE
Add crypto lock

### PR DESCRIPTION
**Pull request by hieronymus777**

### CODE REVIEW

**Review by hishbase — CHANGES_REQUESTED** [review]

**Review by hishbase — CHANGES_REQUESTED** [review]

**Review by hishbase — APPROVED** [review]

**Review by hishbase — CHANGES_REQUESTED** [review]
Maybe we should have two separate methods for `getSecretKey`. Maybe call it `getAES256GMSecretKey` vs whatever the app is using. The idea is we're passing `alias` which is   is also used in to construct `spec` param outside. We don't want these to be different. 

Let chat about this real quick on Monday

**Comment by alexshoykhet** [comment]
Updated the pr per our conversation.  Thanks!

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+import kotlin.concurrent.withLock
+
+object KeyStores {
+    /**
+     * Get a key spec for gcm block mode without requiring user authentication
+     *
+     * @param alias the alias of the key
+     * @return a key spec for gcm block mode without user authentication
+     */
+    fun getGCMKeySpec(alias: String): KeyGenParameterSpec = getKeyGenParameterSpecBuilder(alias).build()
```
Comment (hishbase) [inline android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: nit: `buildGCMKeySpec`?

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+
+    /**
+     * Get a key spec for gcm block mode and require user authentication.
+     *
+     * @param alias the alias of the key
+     * @return a key spec for gcm block mode with user authentication
+     */
+    fun getGCMWithUserAuthenticationKeySpec(alias: String): KeyGenParameterSpec = getKeyGenParameterSpecBuilder(alias)
```
Comment (hishbase) [inline android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: nit: merge with `buildGCMKeySpec` above and just pass a parameter i.e.

```kotlin
fun buildGCMKeySpec(alias: String, isAuthenticated: Boolean = false)
```

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+     * @param keystore the keystore to delete the alias from
+     * @param alias the alias of the key to delete
+     */
+    fun delete(keystore: String, alias: String) = CipherLock.withLock {
+        val keyStore = KeyStore.getInstance(keystore)
+        keyStore.load(null)
+        if (keyStore.containsAlias(alias)) keyStore.deleteEntry(alias)
+    }
+
+    private fun getKeyGenParameterSpecBuilder(alias: String): KeyGenParameterSpec.Builder = KeyGenParameterSpec
```
Comment (hishbase) [inline android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: Based on the above comments, you can move this to the body of `buildGCMKeySpec`

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+     * @return an AES [SecretKey]
+     */
+    @Throws(
+        KeyStoreException::class,
+        IOException::class,
+        NoSuchAlgorithmException::class,
+        CertificateException::class,
+        UnrecoverableEntryException::class
+    )
+    fun getAESWithGCMSecretKey(keystore: String, alias: String, isAuthenticated: Boolean = false): SecretKey {
```
Comment (hishbase) [inline android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: nit: `getOrCreateAES256GCMSecretKey`

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+     * @return an AES [SecretKey]
+     */
+    @Throws(
+        KeyStoreException::class,
+        IOException::class,
+        NoSuchAlgorithmException::class,
+        CertificateException::class,
+        UnrecoverableEntryException::class
+    )
+    fun getAESSecretKey(keystore: String, spec: KeyGenParameterSpec): SecretKey = CipherLock.withLock {
```
Comment (hishbase) [inline android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: nit: `getOrCreateAESSecretKey`

**Review by hishbase — CHANGES_REQUESTED** [review]

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt (hunk 5 above)
Comment (alexshoykhet) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: 3bc1ba8d443523dafcce732a1d79390218c1bb1d

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt (hunk 4 above)
Comment (alexshoykhet) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: 3bc1ba8d443523dafcce732a1d79390218c1bb1d

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt (hunk 3 above)
Comment (alexshoykhet) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: 3bc1ba8d443523dafcce732a1d79390218c1bb1d

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt (hunk 2 above)
Comment (alexshoykhet) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: 3bc1ba8d443523dafcce732a1d79390218c1bb1d

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt (hunk 1 above)
Comment (alexshoykhet) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/KeyStores.kt]: 3bc1ba8d443523dafcce732a1d79390218c1bb1d

**Review by hishbase — APPROVED** [review]

File: android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/AES256GCM.kt [hunk 6]
```diff
@@ -68,15 +78,20 @@ class AES256GCM {
          * @return The decrypted data
          * @throws`EncryptionError.invalidAES256GCMData` if unable to decrypt data
          */
-        @Synchronized fun decrypt(data: ByteArray, key: ByteArray, iv: ByteArray, authTag: ByteArray): ByteArray {
-            val cipher = Cipher.getInstance(TRANSFORMATION)
-            val paramSpec = GCMParameterSpec(AUTH_TAG_SIZE, iv)
-            val keySpec = SecretKeySpec(key, "AES")
-            val encryptedData = data + authTag
+        fun decrypt(data: ByteArray, key: ByteArray, iv: ByteArray, authTag: ByteArray): ByteArray {
```
[the remark this answers is not in the document]
Comment (tuanggolt) [reply android/crypto/src/main/java/com/coinbase/wallet/crypto/ciphers/AES256GCM.kt]: > Contributor

**Review by tuanggo — COMMENTED** [review]
open

**Pull request merged**